### ISSUE TITLE
[🚀 Feature]: Selenium Manager downloads fixed version of Chrome browser (Rust)

### ISSUE THREAD

**Issue by bonigarcia** [SeleniumHQ/selenium#11678]
### Feature and motivation

The next major feature to be implemented in Selenium Manager is the automated management (download, cache) of the principal browsers nowadays: Chrome, Firefox, and Edge.

This issue gets track of the implementation of this feature for managing a given version of **Chrome** (e.g., 101) on the Rust side.

### Usage example

```
selenium-manager --browser chrome --browser-version 101 --resolve-browser
```

**Comment by diemol** [comment]
I missed this issue before. The problem with Chrome is that you cannot download old versions. You can only download the current, beta, dev, and canary.

**Comment by bonigarcia** [comment]
That's true. The alternative to install former versions of Chrome is installing [Chromium](https://www.chromium.org/getting-involved/download-chromium/).

**Comment by titusfortner** [comment]
ooc, How different is Chrome from Chromium?

**Comment by bonigarcia** [comment]
A Chromium build is equivalent to Chrome but doesn't include proprietary parts, such as proprietary media formats such as AAC, H.264, or MP3. Also, Chrome is evergreen (i.e., it updates itself automatically), while Chromium is not. You can check a detailed comparison [here](https://www.browserstack.com/guide/difference-between-chrome-and-chromium) or [here](https://www.lambdatest.com/blog/difference-between-chrome-and-chromium/).

I would say that for most automated testing use cases, Chromium and Chrome are equivalent. Of course, there can be some cases in which Chrome (and not Chromium) is mandatory (e.g., videoconferencing web applications), but I believe it is a minority.

**Comment by titusfortner** [comment]
Do we even have a good place to download chromium binaries from?

I think we're ok using equivalent Chromium if people want historical versions

**Comment by bonigarcia** [comment]
> Do we even have a good place to download chromium binaries from?

Yes, it is explained in the link I posted above ([Chromium](https://www.chromium.org/getting-involved/download-chromium/), see "Downloading old builds of Chrome / Chromium"). All the Chromium releases are hosted in the official [repository](https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html). The idea is that Selenium Manager will do that process in an automated manner when the user requests an old version of Chrome.

**Comment by titusfortner** [comment]
Right, but we need to be able to match up the version number with the build number, and I haven't seen anywhere that maintains a good list of that; would we need to create our own?

**Comment by bonigarcia** [comment]
I've been looking into the procedure published on the previous page to get old versions of Chromium. The tool they use to track Chromium versions is called [OmahaProxy](https://omahaproxy.appspot.com/), although this tool will be turned down on April 14, 2023, and they recommend [ChromiumDash](https://chromiumdash.appspot.com/) instead.

Supposing we want to download a given version of Chrome (e.g., Chromium 110 for Windows), the first step is to find the exact release version. ChromiumDash has a page about releases to track this info:

https://chromiumdash.appspot.com/releases?platform=Windows

With that version (in this example, 110.0.5481.179), we need to find a "base position" to download a Chromium binary (for Windows, Linux, or Mac). For that, we can make a query to the following page:

https://omahaproxy.appspot.com/deps.json?version=110.0.5481.179

The response to that request is as follows:

```
{"chromium_version": "110.0.5481.179", "skia_commit": "aab9fb4100da797d25fe340e9a2fcb2ae30fc2e1", "chromium_base_position": "1084008", "v8_version": "11.0.226.16", "v8_commit": "7999223ca1644726339aae43d9435c721c8a4bb0", "chromium_branch": "5481", "v8_position": "32", "chromium_base_commit": "130f3e4d850f4bc7387cfb8d08aa993d288a67a9", "chromium_commit": "54da1f7102fa09a25e8603c5e509117c91fe4930"}
```

In light of the response, the base position we are looking for is `1084008`. In theory, we can use that base position to query the snapshot repository using an URL as follows:

https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Win/base-position/

Nevertheless, that base position is not usually exact, and it is explained in the info:

_Sometimes you may have to decrement the commit number until you find one._

For this example, the lowest commit number with a release is `1083984`.

With that number, we can go to the repo:

https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Win/1083984/

... and download `chrome-win.zip`, whose URL is:

https://www.googleapis.com/download/storage/v1/b/chromium-browser-snapshots/o/Win%2F1083984%2Fchrome-win.zip?generation=1671147045082306&alt=media

After unzipping the file, the Chromium browser can be executed:

![image](https://user-images.githubusercontent.com/5413013/231827438-d8b2408a-c8e9-41b7-8d9c-a6d9f12b3cca.png)

In summary, this process was several manual steps. And although it is possible to make this process in a fully automated manner with some tricks, maybe it deserves to maintain our own metadata, mapping the old versions of Chromium (starting with a version we decide) with the precise base position and even the URL in which the release can be downloaded. Selenium Manager simply would request this metadata to find out the URLs in which the Chromium releases are hosted.

**Comment by titusfortner** [comment]
yeah, that's a lot. I agree a basic map would be much easier. 😄

**Comment by diemol** [comment]
This is already implemented, right?

**Comment by bonigarcia** [comment]
Yes, but so far, we are using CfT starting in Chrome 113, the first one CfT supports.

We need to decide if we want to support previous Chrome versions through the Chromium releases, as initially planned (before the advent of CfT). This is still possible (see comments above), making a mapping of the existing Chromium releases before version 113.

**Comment by diemol** [comment]
It doesn't make sense to use Chromium now given that CfT is out there. If someone wants to use Chromium for testing, they must get the binary themselves. We need to keep a scope that helps us have something easy to maintain.

**Comment by bonigarcia** [comment]
That's a decision we need to take. I'm just saying that it is still possible if we want to support downloading older versions of Chrome (i.e., Chromium < 113). In that case, we would need to map the existing Chromium releases (see [dashboard](https://chromiumdash.appspot.com/schedule)) to download Chromium releases (for the different operating systems) easily. In theory, when the mapping is done, no maintenance will be required since that mapping won't change because it is based on older versions.

**Comment by titusfortner** [comment]
I think we should keep this issue open. It's valid but very low priority. We can see if people ask for it.
We know Sauce users test on old versions of browsers all the time.

**Comment by titusfortner** [comment]
I'm voting "no" on this now. 113 and later should be sufficient

**Comment by diemol** [comment]
I agree. CfT is giving us enough.

**Comment by bonigarcia** [comment]
Ok, I'm closing this then.
